src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::context::{self, Context};
+use crate::events::EventNotifier;
+use crate::worker::{self, Worker};
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Order {
+    LeftToRight,
+    RightToLeft,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Strategy {
+    OneForOne,
+}
+
+#[derive(Debug)]
+pub struct StartError {
+    pub failing_child_runtime_name: String,
+    pub failing_child_error: Arc<worker::StartError>,
+    pub failing_sibling_termination_error: Option<TerminationError>,
+}
+
+#[derive(Debug)]
+pub struct TerminationError(HashMap<String, Arc<worker::TerminationError>>);
+
+pub struct Spec {
+    pub name: String,
+    pub ev_notifier: EventNotifier,
+    pub order: Order,
+    pub strategy: Strategy,
+    pub children: Vec<worker::Spec>,
+}
+
+pub struct Supervisor {
+    pub runtime_name: String,
+    pub children: Vec<Worker>,
+
+    pub name: String,
+    pub ev_notifier: EventNotifier,
+    pub order: Order,
+    pub strategy: Strategy,
+}
+
+fn ordered_children_on_start(
+    order: &Order,
+    children_specs: Vec<worker::Spec>,
+) -> Box<dyn Iterator<Item = worker::Spec>> {
+    match order {
+        Order::LeftToRight => Box::new(children_specs.into_iter()),
+        Order::RightToLeft => Box::new(children_specs.into_iter().rev()),
+    }
+}
+
+fn ordered_children_on_termination(
+    order: &Order,
+    children: Vec<Worker>,
+) -> Box<dyn Iterator<Item = Worker>> {
+    match order {
+        Order::LeftToRight => Box::new(children.into_iter().rev()),
+        Order::RightToLeft => Box::new(children.into_iter()),
+    }
+}
```

Comment:
This is the exact same function as `ordered_children_on_start` and both should be abstracted to:
```suggestion
fn ordered_children<Child>(
    order: Order,
    children: Vec<Child>,
) -> Box<dyn Iterator<Item = Child>> {
    match order {
        Order::LeftToRight => Box::new(children.into_iter().rev()),
        Order::RightToLeft => Box::new(children.into_iter()),
    }
}
```

Reply:
Not quite, note the `.rev` call is on `LeftToRight` on `ordered_children_on_termination`, and in `ordered_children_on_start` the `.rev` call is on the `RightToLeft` arm. 

Maybe we can re-implement these as methods on the `Order` type that sort an input children list with `start` and `stop` semantics? 

```rust
// Not a fan of the Procedure name
#[derive(Copy, Clone, PartialEq, Debug)]
pub enum Procedure = {
  Start
  Termination
} 

impl Order {
  pub fn sort<a>(&self, procedure: Procedure, children: Vec<a>) -> Box<dyn Iterator<Item = a>> {
     let should_revert = match (self, procedure) {
       (Order::LeftToRight, Start) => false,
       (Order::LeftToRight, Stop) => true,
       (Order::RightToLeft, Start) => true,
       (Order::RightToLeft, Stop) => false,
     } ;
    if should_revert {
      Box::new(iterator.rev())
    } else {
      Box::new(iterator)
    }
}
```

Reply:
Oh, good point! I wasn't paying close enough attention! :laughing:

I don't know if this very minor duplication is worth introducing a new enum type over. I think since they are different methods it's fine to leave them as is. Though your new `sort` method is quite nice for collecting the logic into one place so it won't be refactored by someone paying as little attention as I was...

Minor side note though: you could clean up the implementation of `sort` in your proposal by using an `|` to only have two branches to the match statement, and then putting the iterator creation directly in the match, like so:

```rust
// Maybe Lifecycle is better?
#[derive(Copy, Clone, PartialEq, Debug)]
pub enum Lifecycle = {
  Start
  Termination
} 

impl Order {
  pub fn sort<a>(&self, lifecycle: Lifecycle, children: Vec<a>) -> Box<dyn Iterator<Item = a>> {
     match (self, procedure) {
       (Order::LeftToRight, Start) | (Order::RightToLeft, Termination) => Box::new(iterator),
       (Order::LeftToRight, Termination) | (Order::RightToLeft, Start) => Box::new(iterator.rev()),
     }
}
```